numpy/f2py/cfuncs.py

```diff
@@ -481,17 +481,26 @@
 """
 needs['STRINGCOPYN'] = ['string.h', 'FAILNULL']
 cppmacros['STRINGCOPYN'] = """\
-#define STRINGCOPYN(to,from,buf_size)                           \\
+/*
+STRINGCOPYN copies N bytes.
+
+`to` and `from` buffers must have sizes of at least N bytes.
+
+If the last byte in `to` is null (`\\0`), then replace all the
+preceeding nulls with spaces.
```

Comment:
Why do we not replace the trailing null as well?

Reply:
This is the current behavior:
* `STRINGCOPYN(to, b"a", 4) = b"a  \0"
* `STRINGCOPYN(to, b"ab", 4) = b"ab \0"
* `STRINGCOPYN(to, b"abc", 4) = b"abc\0"
* `STRINGCOPYN(to, b"abcd", 4) = b"abcd"

"Add a `\0` if there's room otherwise don't" smells like a bug to me

Reply:
Notice that `\0` comes from the input and is not added by `STRINGCOPYN`.

The bug is in usage, not in `STRINGCOPYN`: the value N=4 violates the contract:
```
`to` and `from` buffers must have sizes of at least N bytes.
```
for the first and second example.

Reply:
If that is actually the contract then there is no point using `strncopy` as a simple `memcpy` would have the same effect

Reply:
> Why do we not replace the trailing null as well?

Because the trailing null is not a part of the fixed-width string.

Reply:
> If that is actually the contract then there is no point using `strncopy` as a simple `memcpy` would have the same effect

`strncpy` does not access `from` after the first appearance of `\0` value. `memcpy` does. So, when using `memcpy`, the example `STRINGCOPYN(to, b"a\0", 4)` would have memory violation while with `strncpy` not.
Although, within the f2py generated code, this situation should never happen.

Reply:
> Because the trailing null is not a part of the fixed-width string.

It sounds like there's an implicit contract I'm missing then - is `STRINGCOPYN` intended for converting fixed-width strings into null-terminated ones, or for going in the other direction?

Reply:
`STRINGCOPYN` is used in both directions:
- in `string_from_pyobj` to copy (possibly shorter) C string to fixed-width Fortran string
- in `try_pyarr_from_string` to copy fixed-width Fortran (sub)string to (possibly shorter) C string.